pkg/controller/cloneset/cloneset_controller.go

```diff
@@ -536,3 +577,132 @@ func (r *ReconcileCloneSet) claimPods(instance *appsv1alpha1.CloneSet, pods []*v
 
 	return claimedPods, nil
 }
+
+func (r *ReconcileCloneSet) createImagePullJobsForInPlaceUpdate(cs *appsv1alpha1.CloneSet, currentRevision, updateRevision *apps.ControllerRevision) error {
+	if _, ok := updateRevision.Labels[appsv1alpha1.ImagePreDownloadCreatedKey]; ok {
+		return nil
+	} else if _, ok := updateRevision.Labels[appsv1alpha1.ImagePreDownloadIgnoredKey]; ok {
+		return nil
+	}
+
+	// ignore if update type is ReCreate
+	if cs.Spec.UpdateStrategy.Type == appsv1alpha1.RecreateCloneSetUpdateStrategyType {
+		klog.V(4).Infof("CloneSet %s/%s skipped to create ImagePullJob for update type is %s",
+			cs.Namespace, cs.Name, cs.Spec.UpdateStrategy.Type)
+		return r.patchControllerRevisionLabels(updateRevision, appsv1alpha1.ImagePreDownloadIgnoredKey, "true")
+	}
+
+	// ignore if replicas <= minimumReplicasToPreDownloadImage
+	if *cs.Spec.Replicas <= minimumReplicasToPreDownloadImage {
+		klog.V(4).Infof("CloneSet %s/%s skipped to create ImagePullJob for replicas %d <= %d",
+			cs.Namespace, cs.Name, *cs.Spec.Replicas, minimumReplicasToPreDownloadImage)
+		return r.patchControllerRevisionLabels(updateRevision, appsv1alpha1.ImagePreDownloadIgnoredKey, "true")
+	}
+
+	// ignore if this revision can not update in-place
+	opts := &inplaceupdate.UpdateOptions{}
+	inplaceupdate.SetOptionsDefaults(opts)
+	spec := opts.CalculateSpec(currentRevision, updateRevision, opts)
+	if spec == nil {
+		klog.V(4).Infof("CloneSet %s/%s skipped to create ImagePullJob for %s -> %s can not update in-place",
+			cs.Namespace, cs.Name, currentRevision.Name, updateRevision.Name)
+		return r.patchControllerRevisionLabels(updateRevision, appsv1alpha1.ImagePreDownloadIgnoredKey, "true")
+	}
+
+	// start to create jobs
+
+	var pullSecrets []string
+	for _, s := range cs.Spec.Template.Spec.ImagePullSecrets {
+		pullSecrets = append(pullSecrets, s.Name)
+	}
+
+	selector := cs.Spec.Selector.DeepCopy()
+	selector.MatchExpressions = append(selector.MatchExpressions, metav1.LabelSelectorRequirement{
+		Key:      apps.ControllerRevisionHashLabelKey,
+		Operator: metav1.LabelSelectorOpNotIn,
+		Values:   []string{updateRevision.Name, updateRevision.Labels[history.ControllerRevisionHashLabel]},
+	})
+
+	var pullTimeoutSeconds int32 = 300
+	if str, ok := cs.Annotations[appsv1alpha1.ImagePreDownloadTimeoutSecondsKey]; ok {
+		if i, err := strconv.Atoi(str); err == nil {
+			pullTimeoutSeconds = int32(i)
+		}
+	}
+
+	parallelism := intstr.FromInt(1)
+	if str, ok := cs.Annotations[appsv1alpha1.ImagePreDownloadParallelismKey]; ok {
+		if i, err := strconv.Atoi(str); err == nil {
+			parallelism = intstr.FromInt(i)
+		}
+	}
+
+	labelMap := make(map[string]string)
+	for k, v := range cs.Spec.Template.Labels {
+		labelMap[k] = v
+	}
+	labelMap[history.ControllerRevisionHashLabel] = updateRevision.Labels[history.ControllerRevisionHashLabel]
+
+	for name, image := range spec.ContainerImages {
```

Comment:
can init containers be supported too ?

Reply:
we have not supported to in-place update init containers